lib/babe/babe.go

```diff
@@ -337,74 +339,73 @@ func (b *Service) runEngine() error {
 	}
 
 	for {
-		err = func() error {
-			ctx, cancel := context.WithCancel(context.Background())
-			defer cancel()
-			b.epochHandler, err = b.initiateAndGetEpochHandler(epoch)
-			if err != nil {
-				return err
+		if err = b.handleEpoch(epoch); err != nil {
+			if errors.Is(err, errServicePaused) || errors.Is(err, context.Canceled) {
+				return nil
 			}
 
-			// get start slot for current epoch
-			nextEpochStart, err := b.epochState.GetStartSlotForEpoch(epoch + 1)
-			if err != nil {
-				return fmt.Errorf("failed to get start slot for next epoch %d: %s", epoch+1, err)
-			}
+			return err
+		}
+	}
+}
 
-			nextEpochStartTime := getSlotStartTime(nextEpochStart, b.constants.slotDuration)
-			epochTimer := time.NewTimer(time.Until(nextEpochStartTime))
-			cleanup := func() {
-				if !epochTimer.Stop() {
-					<-epochTimer.C
-				}
-			}
+func (b *Service) handleEpoch(epoch uint64) (err error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	b.epochHandler, err = b.initiateAndGetEpochHandler(epoch)
+	if err != nil {
+		return err
+	}
 
-			errCh := make(chan error)
-			go b.epochHandler.run(ctx, errCh)
+	// get start slot for current epoch
+	nextEpochStart, err := b.epochState.GetStartSlotForEpoch(epoch + 1)
+	if err != nil {
+		return fmt.Errorf("failed to get start slot for next epoch %d: %w", epoch+1, err)
+	}
 
-			select {
-			case <-b.ctx.Done():
-				cleanup()
-				return b.ctx.Err()
-			case <-b.pause:
-				cleanup()
-				return errServicePaused
-			case <-epochTimer.C:
-				// stop current epoch handler
-				cancel()
-			case err := <-errCh:
-				logger.Errorf("error from epochHandler: %s", err)
-			}
+	nextEpochStartTime := getSlotStartTime(nextEpochStart, b.constants.slotDuration)
+	epochTimer := time.NewTimer(time.Until(nextEpochStartTime))
+	cleanup := func() {
+		if !epochTimer.Stop() {
+			<-epochTimer.C
+		}
+	}
 
-			// setup next epoch, re-invoke block authoring
-			next, err := b.incrementEpoch()
-			if err != nil {
-				return fmt.Errorf("failed to increment epoch: %s", err)
-			}
+	errCh := make(chan error)
+	go b.epochHandler.run(ctx, errCh)
 
-			logger.Infof("epoch %d complete, upcoming epoch: %d", epoch, next)
-			epoch = next
-			return nil
-		}()
-		if err != nil {
-			if errors.Is(err, errServicePaused) || errors.Is(err, context.Canceled) {
-				return nil
-			}
+	select {
+	case <-b.ctx.Done():
+		cleanup()
+		return b.ctx.Err()
+	case <-b.pause:
+		cleanup()
+		return errServicePaused
+	case <-epochTimer.C:
+		// stop current epoch handler
+		cancel()
+	case err := <-errCh:
```

Comment:
You should also call `cleanup()` here
```suggestion
	case err := <-errCh:
		cleanup()
```